### PR TITLE
Reverse geocoding

### PR DESCRIPTION
**Pull request by CarteKiwi**
Reverse Geocoding now added to the project.

Call **ReverseGeocodeCurrentLocation()** or **ReverseGeocodeLocation(lat, long)** to get an address.

### CODE REVIEW

File: src/Geolocator.Plugin.WindowsPhone8/GeolocatorImplementation.cs [hunk 1]
```diff
@@ -113,6 +113,26 @@ public Task<Position> GetPositionAsync(TimeSpan? timeout, CancellationToken? can
         }
 
         /// <summary>
+        /// Gets position async and reverse geocode
+        /// </summary>
+        /// <returns>Address of the current position</returns>
+        public Task<Address> ReverseGeocodeCurrentLocation()
+        {
+            throw new NotImplementedException();
```
Comment (jamesmontemagno) [inline src/Geolocator.Plugin.WindowsPhone8/GeolocatorImplementation.cs]: Should this throw? or just return null?

File: src/Geolocator.Plugin.WindowsPhone81/GeolocatorImplementation.cs [hunk 2]
```diff
@@ -172,6 +172,26 @@ public Task<Position> GetPositionAsync(TimeSpan? timeout, CancellationToken? can
         }
 
         /// <summary>
+        /// Gets position async and reverse geocode
+        /// </summary>
+        /// <returns>Address of the current position</returns>
+        public Task<Address> ReverseGeocodeCurrentLocation()
+        {
+            throw new NotImplementedException();
```
Comment (jamesmontemagno) [inline src/Geolocator.Plugin.WindowsPhone81/GeolocatorImplementation.cs]: Same as above, I assume there has to be an implementation somewhere in Xamarin.Forms world?

Comment (jamesmontemagno) [inline src/Geolocator.Plugin.WindowsPhone81/GeolocatorImplementation.cs]: Maybe this: https://github.com/xamarin/Xamarin.Forms/blob/master/Xamarin.Forms.Maps.UWP/GeocoderBackend.cs ?

File: src/Geolocator.Plugin.WindowsPhone8/GeolocatorImplementation.cs (hunk 1 above)
Comment (jamesmontemagno) [inline src/Geolocator.Plugin.WindowsPhone8/GeolocatorImplementation.cs]: Should be able to utilize: https://github.com/xamarin/Xamarin.Forms/blob/master/Xamarin.Forms.Maps.WP8/GeocoderBackend.cs

File: src/Geolocator.Plugin.Abstractions/Address.cs [hunk 3]
```diff
@@ -0,0 +1,70 @@
+﻿
+//
```
Comment (jamesmontemagno) [inline src/Geolocator.Plugin.Abstractions/Address.cs]: We can remove this license up top

File: src/Geolocator.Plugin.Android/GeolocatorImplementation.cs [hunk 4]
```diff
@@ -239,6 +236,36 @@ public double DesiredAccuracy
...
 
+        /// <summary>
+        /// Gets position async and reverse geocode
+        /// </summary>
+        /// <returns>Address of the current position</returns>
+        public async Task<Address> ReverseGeocodeCurrentLocation()
+        {
+            var pos = await GetPositionAsync(new TimeSpan(0, 0, 1));
+
+            Geocoder geocoder = new Geocoder(Application.Context);
```
Comment (jamesmontemagno) [inline src/Geolocator.Plugin.Android/GeolocatorImplementation.cs]: Can you make it "var" geocoder

File: src/Geolocator.Plugin.Android/GeolocatorImplementation.cs [hunk 5]
```diff
@@ -239,6 +236,36 @@ public double DesiredAccuracy
...
+
+        /// <summary>
+        /// Reverse geocode a position
+        /// </summary>
+        /// <param name="latitude">Desired Latitude</param>
+        /// <param name="longitude">Desired Longitude</param>
+        /// <returns>Address of the desired position</returns>
+        public async Task<Address> ReverseGeocodeLocation(double latitude, double longitude)
+        {
+            Geocoder geocoder = new Geocoder(Application.Context);
```
Comment (jamesmontemagno) [inline src/Geolocator.Plugin.Android/GeolocatorImplementation.cs]: Can you make it "var" geocoder

File: src/Geolocator.Plugin.Abstractions/IGeolocator.cs [hunk 6]
```diff
@@ -62,14 +62,28 @@ public interface IGeolocator
...
+        /// <returns>Address of the current position</returns>
+        Task<Address> ReverseGeocodeCurrentLocation();
+
+        /// <summary>
+        /// Reverse geocode a position
+        /// </summary>
+        /// <param name="latitude">Desired Latitude</param>
+        /// <param name="longitude">Desired Longitude</param>
+        /// <returns>Address of the desired position</returns>
+        Task<Address> ReverseGeocodeLocation(double latitude, double longitude);
```
Comment (jamesmontemagno) [inline src/Geolocator.Plugin.Abstractions/IGeolocator.cs]: Do you think this should be "IEnumerable<Address>"? That way we don't have to do first or default. Like Xamarin.Forms does.

File: src/Geolocator.Plugin.Abstractions/IGeolocator.cs [hunk 7]
```diff
@@ -62,14 +62,28 @@ public interface IGeolocator
...
-		/// <param name="minimumTime">Time</param>
-		/// <param name="minimumDistance">Distance</param>
-		/// <param name="includeHeading">Include heading or not</param>
-		/// <param name="listenerSettings">Optional settings (iOS only)</param>
-		Task<bool> StartListeningAsync(TimeSpan minimumTime, double minimumDistance, bool includeHeading = false, ListenerSettings listenerSettings = null);
+        /// <summary>
+        /// Gets position async and reverse geocode
+        /// </summary>
+        /// <returns>Address of the current position</returns>
+        Task<Address> ReverseGeocodeCurrentLocation();
```
Comment (jamesmontemagno) [inline src/Geolocator.Plugin.Abstractions/IGeolocator.cs]: I don't think we should combine the two of them together at all. I say make them pass it in.

File: src/Geolocator.Plugin.WindowsPhone8/GeolocatorImplementation.cs (hunk 1 above)
Comment (CarteKiwi) [inline src/Geolocator.Plugin.WindowsPhone8/GeolocatorImplementation.cs]: Done

File: src/Geolocator.Plugin.WindowsPhone81/GeolocatorImplementation.cs (hunk 2 above)
Comment (CarteKiwi) [inline src/Geolocator.Plugin.WindowsPhone81/GeolocatorImplementation.cs]: Done

File: src/Geolocator.Plugin.Android/GeolocatorImplementation.cs (hunk 4 above)
Comment (CarteKiwi) [inline src/Geolocator.Plugin.Android/GeolocatorImplementation.cs]: Done

File: src/Geolocator.Plugin.Android/GeolocatorImplementation.cs (hunk 5 above)
Comment (CarteKiwi) [inline src/Geolocator.Plugin.Android/GeolocatorImplementation.cs]: Done

File: src/Geolocator.Plugin.Abstractions/IGeolocator.cs (hunk 7 above)
Comment (CarteKiwi) [inline src/Geolocator.Plugin.Abstractions/IGeolocator.cs]: Removed

File: src/Geolocator.Plugin.Abstractions/IGeolocator.cs (hunk 6 above)
Comment (CarteKiwi) [inline src/Geolocator.Plugin.Abstractions/IGeolocator.cs]: Done

File: src/Geolocator.Plugin.Abstractions/Address.cs (hunk 3 above)
Comment (CarteKiwi) [inline src/Geolocator.Plugin.Abstractions/Address.cs]: Done

**Comment by CarteKiwi** [comment]
I'm glad you appreciate it. Thank you for your enthusiasm :)
I have followed your advices. 
But, I have not implemented W8.1 because I have no solution right now. Have to investigate a little more but I'm running out of time today :)

**Comment by jamesmontemagno** [comment]
Makes sense to me about W81.

Can you add in some docs for it and how it works in the readme? Do a new PR for it, I'll pull this in.

**Comment by alanspires** [comment]
Looking forward to this PR guys.

**Comment by jamesmontemagno** [comment]
You can add the development nuget feed: https://ci.appveyor.com/nuget/geolocatorplugin :)

**Comment by kevincraig** [comment]
Is this available in any of the beta versions yet?  I was getting ready to write my own then saw this Just curious.  Thanks.  KC

**Comment by jamesmontemagno** [comment]
It is in the beta

**Pull request merged**